lib/Dialect/SV/Transforms/AlwaysFusion.cpp

```diff
@@ -0,0 +1,131 @@
+static void mergeRegions(Region *region1, Region *region2) {
+  assert(region1->getBlocks().size() <= 1 && region2->getBlocks().size() <= 1 &&
+         "Can only merge regions with a single block");
+  if (region1->empty()) {
+    // If both regions are empty, move on to the next pair of regions
+    if (region2->empty())
+      return;
+    // If the first region has no block, move the second region's block over.
+    region1->getBlocks().splice(region1->end(), region2->getBlocks());
+    return;
+  }
+  // If the second region is not empty, splice its block into the end of the
+  // first region.
+  if (!region2->empty()) {
+    auto &block1 = region1->front();
+    auto &block2 = region2->front();
+    // Remove the terminator from the first block before merging.
+    block1.back().erase();
+    block1.getOperations().splice(block1.end(), block2.getOperations());
+  }
+}
+
+/// Inline all regions from the second operation into the first.
+static void mergeOperations(Operation *op1, Operation *op2) {
+  assert(op1 != op2 && "Cannot merge an op into itself");
+  for (unsigned i = 0; i < op1->getNumRegions(); ++i)
+    mergeRegions(&op1->getRegion(i), &op2->getRegion(i));
+}
+
+struct AlwaysFusionPass : public AlwaysFusionBase<AlwaysFusionPass> {
+  void runOnOperation() override {
+    auto *op = getOperation();
+
+    // Keeps track if anything changed during this pass, used to determine if
+    // the analyses were preserved.
+    bool graphChanged = false;
+
+    auto kindInterface = dyn_cast<mlir::RegionKindInterface>(op);
+    for (unsigned i = 0; i < op->getNumRegions(); ++i) {
+      // If the operation does not implement the region kind interface, all of
+      // its regions are implicitly regular SSACFG region. Since we are blindly
+      // combining alwaysff blocks, make sure they are in a graph region.
+      if (!kindInterface ||
+          kindInterface.getRegionKind(i) == RegionKind::SSACFG)
+        continue;
+
+      if (op->getRegion(i).empty())
+        continue;
+
+      // Graph regions only have 1 block.
+      auto &block = op->getRegion(i).front();
+
+      // A set of operations in the current block which are mergable. Any
+      // operation in this set is a candidate for another similar operation to
+      // merge in to.
+      DenseSet<Operation *, SimpleOperationInfo> foundOps;
+      for (sv::AlwaysFFOp alwaysOp :
+           llvm::make_early_inc_range(block.getOps<sv::AlwaysFFOp>())) {
+        // check if we have encountered an equivalent operation already.  If we
+        // have, merge them together and delete the old operation.
+        auto it = foundOps.find(alwaysOp);
+        if (it != foundOps.end()) {
+          mergeOperations(*it, alwaysOp);
+          alwaysOp.erase();
+          graphChanged = true;
+          continue;
+        }
+        // Record new mergable ops as a candidate for merging
+        foundOps.insert(alwaysOp);
```

Comment:
alwaysOp will might have just been deleted, yes?  shouldn't this be in an else block?

Reply:
Switched the code around to use an if/else instead of the continue